advanced_source/ddp_pipeline_tutorial.py

```diff
@@ -0,0 +1,455 @@
+`Sequence-to-Sequence Modeling with nn.Transformer and TorchText <https://pytorch.org/tutorials/beginner/transformer_tutorial.html>`__ tutorial
+and scales up the same model to demonstrate how Distributed Data Parallel and
+Pipeline Parallelism can be used to train Transformer models.
+
+Prerequisites:
+
+    * `Pipeline Parallelism <https://pytorch.org/docs/stable/pipeline.html>`__
+    * `Sequence-to-Sequence Modeling with nn.Transformer and TorchText <https://pytorch.org/tutorials/beginner/transformer_tutorial.html>`__
+    * `Getting Started with Distributed Data Parallel <https://pytorch.org/tutorials/intermediate/ddp_tutorial.html>`__
+"""
+
+
+######################################################################
+# Define the model
+# ----------------
+#
+
+
+######################################################################
+# In this tutorial, we will split a Transformer model across two GPUs and use
+# pipeline parallelism to train the model. In addition to this, we use
+# `Distributed Data Parallel <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html>`__
+# to train two replicas of this pipeline. We have one process driving a pipe across
+# GPUs 0 and 1 and another process driving a pipe across GPUs 2 and 3. Both these
+# processes then use Distributed Data Parallel to train the two replicas. The
+# model is exactly the same model used in the `Sequence-to-Sequence Modeling with nn.Transformer and TorchText
+# <https://pytorch.org/tutorials/beginner/transformer_tutorial.html>`__ tutorial,
+# but is split into two stages. The largest number of parameters belong to the
+# `nn.TransformerEncoder <https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html>`__ layer.
+# The `nn.TransformerEncoder <https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html>`__
+# itself consists of ``nlayers`` of `nn.TransformerEncoderLayer <https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html>`__.
+# As a result, our focus is on ``nn.TransformerEncoder`` and we split the model
+# such that half of the ``nn.TransformerEncoderLayer`` are in ``TransformerModelStage1``
+# and the other half are in ``TransformerModelStage2``.
+
+import sys
+import os
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import tempfile
+from torch.nn import TransformerEncoder, TransformerEncoderLayer
+
+if sys.platform == 'win32':
+    print('Windows platform is not supported for pipeline parallelism')
+    sys.exit(0)
+if torch.cuda.device_count() < 4:
+    print('Need at least four GPU devices for this tutorial')
+    sys.exit(0)
+
+class TransformerModelStage1(nn.Module):
+
+    def __init__(self, ntoken, ninp, nhead, nhid, nlayers, dropout=0.5):
+        super(TransformerModelStage1, self).__init__()
+        self.src_mask = None
+        self.pos_encoder = PositionalEncoding(ninp, dropout)
```

Comment:
In the rendered webpage, this is the first block of code and `PositionalEncoding` is defined in the 2nd block. Shall we define it above this so users can copy and paste individual blocks and the script still runs without errors?